examples/multi_task.py

```diff
@@ -0,0 +1,389 @@
+            ranking_labels, pred_rating, sample_weight
+        )
+
+        # Total loss is a weighted combination of the two losses.
+        total_loss = (
+            self.retrieval_loss_wt * retrieval_loss
+            + self.ranking_loss_wt * ranking_loss
+        )
+
+        return total_loss
+
+    def compute_metrics(self, x, y, y_pred, sample_weight=None):
+        # RMSE can be computed irrespective of whether we are
+        # training/evaluating.
+        self.rmse_metric.update_state(
+            y,
+            y_pred["rating"],
+            sample_weight=sample_weight,
+        )
+
+        if "predictions" in y_pred:
+            # We are evaluating or predicting. Update `top_k_metric`.
+            movie_ids = x["movie_id"]
+            predictions = y_pred["predictions"]
+            # For `top_k_metric`, which is a `SparseTopKCategoricalAccuracy`, we
+            # only take top rated movies, and we put a weight of 0 for the rest.
+            rating_weight = keras.ops.cast(keras.ops.greater(y, 0.9), "float32")
+            sample_weight = (
+                rating_weight
+                if sample_weight is None
+                else keras.ops.multiply(rating_weight, sample_weight)
+            )
+            self.top_k_metric.update_state(
+                movie_ids, predictions, sample_weight=sample_weight
+            )
+
+            return self.get_metrics_result()
+        else:
+            # We are training. `top_k_metric` is not updated and is zero, so
+            # don't report it.
+            result = self.get_metrics_result()
+            result.pop(self.top_k_metric.name)
+            return result
+
+
+"""
+## Training and evaluating
+
+We will train three different models here. This can be done easily by passing
+the correct loss weights:
+
+1. Rating-specialised model
+2. Retrieval-specialised model
+3. Multi-task model
+"""
+
+# Rating-specialised model
+model = MultiTaskModel(
+    num_users=users_count + 1,
+    num_candidates=movies_count + 1,
+    ranking_loss_wt=1.0,
+    retrieval_loss_wt=0.0,
+)
+model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))
+model.fit(train_ratings, epochs=5)
+
+model.evaluate(test_ratings)
+
+# Retrieval-specialised model
+model = MultiTaskModel(
+    num_users=users_count + 1,
+    num_candidates=movies_count + 1,
+    ranking_loss_wt=0.0,
+    retrieval_loss_wt=1.0,
+)
+model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))
+model.fit(train_ratings, epochs=5)
+
+model.evaluate(test_ratings)
+
+# Multi-task model
+model = MultiTaskModel(
+    num_users=users_count + 1,
+    num_candidates=movies_count + 1,
+    ranking_loss_wt=1.0,
+    retrieval_loss_wt=1.0,
+)
+model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))
+model.fit(train_ratings, epochs=5)
+
+model.evaluate(test_ratings)
+
+"""
+Let's plot a table of the metrics and pen down our observations:
+
+| Model                 | Top-K Accuracy (↑) | RMSE (↓) |
```

Comment:
Do the little up / down arrows in the table header do anything?

Reply:
It just means that a higher value of accuracy is better, and a lower value of RMSE is better

Reply:
Oh ok